MvvmCross/Navigation/MvxNavigationService.cs

```diff
@@ -256,37 +256,53 @@ public virtual Task<bool> CanNavigate(Type viewModelType)
 
         protected virtual async Task Navigate(MvxViewModelRequest request, IMvxViewModel viewModel, IMvxBundle presentationBundle = null, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var args = new NavigateEventArgs(viewModel);
+            bool hasNavigated = false;
+            if (cancellationToken != default(CancellationToken))
+            {
+                cancellationToken.Register(async () =>
+                {
+                    if(hasNavigated)
+                        await Close(viewModel);
```

Comment:
This is dangerous. We have full async support within the navigation, so this is no longer required. If navigation has completed, it should be the responsibility of the destination page to close and return to previous page, rather than some built in mechanism. This is just going to lead to some nasty race conditions

Reply:
Good point @nickrandolph. Can you elaborate a bit more on why was this added @martijn00?

Reply:
Good point @nickrandolph. Can you elaborate a bit more on why was this added @martijn00?

Reply:
@nmilcoff I think this block of code (repeated in the Navigate method that returns a result) shouldn't be here. It should only be there for Navigations that return a result, so that the caller cancel the navigation. Think of the scenario where you prompt the user for some information by navigating to a page that will return a result - you might then want to cancel this after a period of time. In the case of the Navigate method that returns a result, we should be checking to see if the result has been returned (task completion source has completed) and not navigate if the result has been returned.